dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalBg.cpp

```diff
@@ -317,22 +315,77 @@ bool shouldCompactDeltaWithStable(const DMContext & context, const SegmentSnapsh
     return (delete_rows >= stable_rows * invalid_data_ratio_threshold) || (delete_bytes >= stable_bytes * invalid_data_ratio_threshold);
 }
 
-bool shouldCompactStableWithTooMuchDataOutOfSegmentRange(const SegmentSnapshotPtr & snap, double invalid_data_ratio_threshold, const LoggerPtr & log)
+std::unordered_set<UInt64> getDMFileIDs(const SegmentPtr & seg)
 {
-    auto valid_rows = snap->stable->getRows();
-    auto valid_bytes = snap->stable->getBytes();
+    std::unordered_set<UInt64> file_ids;
+    // We get the file ids in the segment no matter it is abandoned or not,
+    // because even it is abandoned, we don't know whether the new segment will ref the old dtfiles.
+    // So we just be conservative here to return the old file ids.
+    // This is ok because the next check will get the right file ids in such case.
+    if (seg)
+    {
+        const auto & dm_files = seg->getStable()->getDMFiles();
+        for (const auto & file : dm_files)
```

Comment:
BTW I'm curious, when will we have multi DTfiles in the stable currently?

Reply:
Actually there is just one DTFile now. But we keep the interface like this to avoid the assumption that there is only one DTFile for future extenstion.